### ISSUE TITLE
Migrate dflash example scripts to unified CLI

### ISSUE THREAD

**Issue by rahul-tuli** [rahul-tuli/speculators#82]
Type: task

## Question

Migrate the **dflash** family's two example scripts —
`examples/train/dflash_qwen3_8b_sharegpt_online_5k.sh` and
`examples/train/dflash_qwen3_8b_ultrachat_online_5k_bestpractices.sh` — off
the deprecated `scripts/train.py`/`scripts/launch_vllm.py` shims and onto
the unified `speculators` CLI, following the pattern established by
`examples/train/dflash_qwen3_regen_gsm8k_online.sh` (reference only, not
committed — see the map's Notes).

## Definition of done

- Both scripts rewritten to use the unified CLI end-to-end.
- One of the two run end-to-end on this 8xH100 box (screen session, per
  [[feedback-use-screen]]), producing a real checkpoint — this satisfies
  the dflash-family representative-verification requirement (see map
  Notes). Record which script was chosen as the representative and why.
- Resolution comment records: exact commands run, GPU split, and pointers
  to checkpoint output / logs proving completion.

**Comment by rahul-tuli** [comment]
## Resolved

Both dflash scripts migrated to the unified `speculators` CLI and one run end-to-end on this 8xH100 box, producing a real checkpoint.

### Changes
- `examples/train/dflash_qwen3_8b_sharegpt_online_5k.sh`
- `examples/train/dflash_qwen3_8b_ultrachat_online_5k_bestpractices.sh`

Per-script transforms:
- `python scripts/prepare_data.py` → `speculators prepare-data`
- `torchrun … scripts/train.py` → `python -m torch.distributed.run … -m speculators.train`
- `scripts/launch_vllm.py` **kept** — it is a real tool, not a deprecated shim (only `prepare_data.py`/`train.py` carry the `v0.9.0` deprecation warning).
- Added venv-path vars (`SPECULATORS_PY`, `SPECULATORS_BIN`) matching the reference regen script's style.

### Non-obvious finding: step reordering was required (applies to all online scripts)
A pure shim→CLI swap produces a **broken** script. PR #975 (`update preprocess.py to use vllm render endpoint`) changed `prepare-data` so that natural-language datasets (sharegpt/ultrachat) now **require `--render-endpoint`** pointing at a live vLLM server — it derives loss masks from vLLM render boundaries. The old scripts ran prepare-data *before* launching any server, so that ordering is now stale.

Both scripts were reordered to the flow the train tutorial documents:
1. Launch the vLLM server (`launch_vllm.py`, with `--target-layer-ids`).
2. `prepare-data --render-endpoint http://localhost:$VLLM_PORT` against that server.
3. Train against the same server (`--vllm-endpoint …/v1`).

One server serves both the render endpoint (step 2) and hidden-state streaming (step 3). This reorder + `--render-endpoint` is expected to be needed for **every online NL-dataset script** in this migration (dspark/peagle/eagle3-online/mtp).

### Representative verification run
- Chose the **sharegpt** script (the classic recipe; dflash defaults, block_size=8) as the representative.
- The committed scripts keep the general `--data-parallel-size 2` recipe. For the verification run only, used a single vLLM engine (dropped `--data-parallel-size`) per the known box limitation (multi-process DP coordinator fails on this box — see the closed dflash-regen-flow map).
- GPU split: vLLM on GPU 0; training on GPUs 1,2 (`torch.distributed.run --nproc_per_node 2`).
- Data written to `/tmp/dflash_verify_sharegpt` (no fast NVMe on this box).

Exact train invocation (from `train_command.txt`):
```
-m speculators.train --verifier-name-or-path Qwen/Qwen3-8B \
  --data-path /tmp/dflash_verify_sharegpt --vllm-endpoint http://localhost:8000/v1 \
  --save-path /tmp/dflash_verify_sharegpt/checkpoints --draft-vocab-size 32000 \
  --epochs 5 --lr 3e-4 --total-seq-len 8192 --speculator-type dflash \
  --block-size 8 --max-anchors 3072 --num-layers 5 --target-layer-ids 2 18 33 \
  --on-missing generate --on-generate delete
```

Result — pipeline ran end-to-end and produced a real checkpoint:
- prepare-data rendered sharegpt via `/v1/chat/completions/render` and saved 5000 examples (arrow dataset + `token_freq.pt` + `d2t.npy`/`t2d.npy`).
- Epoch 1/5 completed; `train/loss` 1.644 → 1.412, `full_acc` rising.
- Checkpoint at `checkpoints/epoch0_end` → `checkpoints/0/`: `model.safetensors` (2.3 GB), `optimizer_state_dict.pt`, `config.json`, `config.py`, `training_state.json`.

Run was stopped after the first checkpoint (DoD met) to free GPUs; ephemeral `/tmp` output not committed.

Ran in a `screen` session per project convention.

**Comment by rahul-tuli** [comment]
Closing: dflash scripts migrated and representative sharegpt run verified end-to-end with a real checkpoint.
